numerics/unbounded_arrays.hpp

```diff
@@ -90,6 +90,73 @@ class UnboundedLowerTriangularMatrix final {
       UnboundedLowerTriangularMatrix<S> const& matrix);
 };
 
+template<typename Scalar>
+class UnboundedUpperTriangularMatrix final {
+ public:
+  explicit UnboundedUpperTriangularMatrix(int columns);
+  UnboundedUpperTriangularMatrix(int columns, uninitialized_t);
+
+  // The |data| must be in row-major format.
+  UnboundedUpperTriangularMatrix(std::initializer_list<Scalar> const& data);
+
+  void Extend(int extra_columns);
+  void Extend(int extra_columns, uninitialized_t);
+
+  // The |data| must be in row-major format.
+  void Extend(std::initializer_list<Scalar> const& data);
+
+  void EraseToEnd(int begin_column_index);
+
+  int columns() const;
+  int dimension() const;
+
+  bool operator==(UnboundedUpperTriangularMatrix const& right) const;
+
+  // A helper class for indexing column-major data in a human-friendly manner.
+  template<typename Reference>
+  class Row {
+   public:
+    Scalar& operator[](int column);
+    Scalar const& operator[](int column) const;
+
+   private:
+    explicit Row(Reference reference, int row);
+
+    Reference reference_;
```

Comment:
Factor the `&`, and make use the language, rather than the identifiers, to say that it is a reference:

`typename Reference` → `typename QualifiedMatrix`

`Reference reference_` → `QualifiedMatrix& matrix_` (or `that_` if you prefer that king of cleverness)

`Row<UnboundedUpperTriangularMatrix&>` → `Row<UnboundedUpperTriangularMatrix>`
`Row<UnboundedUpperTriangularMatrix const&>` → `Row<UnboundedUpperTriangularMatrix const>`